### ISSUE TITLE
from_numpy_matrix() is ambiguous

### ISSUE THREAD

**Issue by Kuifje02** [networkx/networkx#4876]
Hi, 

I have noticed that the method [networkx.from_numpy_matrix()](https://networkx.org/documentation/stable/reference/generated/networkx.convert_matrix.from_numpy_matrix.html) is ambiguous for weighted graphs, when the graph has an edge with weight 0. If the matrix has a 0 coefficient, then the edge is not created. In otherwords, between two nodes A and B, this method does not make the difference between "there is no edge between A and B" and "there is an edge between A and B, with weight 0". 

On the other hand, the method [networkx.to_numpy_matrix()](https://networkx.org/documentation/stable/reference/generated/networkx.convert_matrix.to_numpy_matrix.html) does make this difference, with the `nonedge` argument.

It would be great to have the same `nonedge` argument for _networkx.from_numpy_matrix()_.

**Comment by rossbar** [comment]
Adding a `nonedge` parameter to `from_numpy_matrix` seems like a reasonable feature request to me. As you allude to, it's straightforward to imagine use cases where the lack of an edge is represented by some other non-zero sentinel value.

**Comment by Kuifje02** [comment]
And the otherway around is very common as well : sometimes you want an edge with 0 cost. For example, if two nodes are superposed geographically, and that the weight represents the distance between the nodes, then obviously the weight is 0, but the edge does exist.

**Comment by jarrodmillman** [comment]
`from_numpy_matrix` is deprecated and will be removed in 3.0.  I am not sure it makes sense to improve a function we are recommending against and will remove in a few months.  `from_numpy_array` has a `nonedge` parameter.

**Comment by rossbar** [comment]
Good point --- thanks Jarrod: all of the above references to `to/from_numpy_matrix` should be replaced by `to/from_numpy_array`.

> from_numpy_array has a nonedge parameter.

Not that I see: https://github.com/networkx/networkx/blob/c73d4ab4e30c5d0dea0aa58e6d22ccb20061fbe3/networkx/convert_matrix.py#L1255

**Comment by dschult** [comment]
It looks like `from_numpy_array` and `from_pandas_adjacency` should both be treated the same way as far as `nonedge`.  We shouldn't need to change `to/from_scipy_sparse_matrix` because 1) they don't have it now, and 2) we don't need a marker to represent "no edge" with a sparse data structure.

Note that this difficulty already breaks `from_numpy_array` when the dtype is not a number type, e.g. `str`. Then all array values are `nonzero` and the result is a complete graph.  It would be nice to have a sentinel string to denote "nonedge" in this case.

I believe the reason it wasn't done before has something to do with the typical troubles with floating point. How do you check if the value is what you think it should be?  The [construct in the current code](https://github.com/networkx/networkx/blob/c73d4ab4e30c5d0dea0aa58e6d22ccb20061fbe3/networkx/convert_matrix.py#L1366) `A.nonzero()` uses a special method for figuring out if a value is zero or not. How do we handle an arbitrary `nonedge`, especially when the dtype of the incoming array is not specified?

The `numpy.nonzero` docs provide the following answer:
```python
edges = ((int(e[0]), int(e[1])) for e in zip(*np.nonzero(A != nonedge)))
```
That might be a place to start.  Something similar should work for pandas.  If you aren't comfortable enough with pandas to tackle both, let's get the numpy version fixed and  then worry about pandas in a separate PR.

The [commented discussion in the code](https://github.com/networkx/networkx/blob/c73d4ab4e30c5d0dea0aa58e6d22ccb20061fbe3/networkx/convert_matrix.py#L1189) warning about trouble and advantages of using `nan` is perhaps of note as well. But I don't think that changes this issue -- its just a warning for people as they pick their sentinel.

**Comment by jarrodmillman** [comment]
> > from_numpy_array has a nonedge parameter.
> 
> Not that I see:
> 
> https://github.com/networkx/networkx/blob/c73d4ab4e30c5d0dea0aa58e6d22ccb20061fbe3/networkx/convert_matrix.py#L1255

Whoops!  I was looking at the wrong function (`to_numpy_arrray`).

**Comment by donn** [comment]
What's a possible workaround here if you want a zero-weighted edge? This is an example; I'm constructing the graph from an adjacency matrix and this was a bit of surprise.

I guess I can add the zero edge manually after the fact…?

```python3
import networkx as nx

G = nx.empty_graph(4)
G.add_edge(0, 1, weight=4)
G.add_edge(0, 2, weight=5)
G.add_edge(0, 3, weight=0)
G.add_edge(1, 2, weight=2)
G.add_edge(1, 3, weight=3)
G.add_edge(2, 3, weight=1)
path = nx.approximation.threshold_accepting_tsp(
    G,
    source=0,
    init_cycle='greedy',
)
print(path)
"[0, 3, 2, 1, 0]"
A = nx.adjacency_matrix(G).todense()
print(A)
"""
[[0 4 5 0]
 [4 0 2 3]
 [5 2 0 1]
 [0 3 1 0]]
"""
Gprime = nx.from_numpy_array(A)
nx.approximation.threshold_accepting_tsp(
    Gprime,
    source=0,
    init_cycle='greedy',
)
"networkx.exception.NetworkXError: G must be a complete graph."
```

**Comment by dschult** [comment]
Currently `nx.from_numpy_array` currently does not support a `nonedge` parameter.

A workaround that ensures creation of a complete graph (without selfloops) involves adding the zero weighted elements manually after the conversion. For example:
```python
## do what you did above
Gprime = nx.from_numpy_array(A)
zero_edges = [e.tolist() for e in np.argwhere(A==0) if e[0]!=e[1]]  # remove the last `if` condition if you want selfloops
Gprime.add_weighted_edges_from((*e, 0) for e in zero_edges)
```
